### PR TITLE
add polkadot.js typesBundle conversion tool

### PR DESCRIPTION
**Pull request by belopash**

### CODE REVIEW

**Review by dzhelezov — CHANGES_REQUESTED** [review]

File: substrate-ingest/src/main.ts [hunk 1]
```diff
@@ -86,7 +88,7 @@ runProgram(async () => {
 
     let typesBundle = options.typesBundle == null
         ? undefined
-        : readOldTypesBundle(options.typesBundle)
+        : readOldTypesBundle(options.typesBundle, options.format)
```
[the remark this answers is not in the document]
Comment (belopash) [reply substrate-ingest/src/main.ts:91]: vut it's validated in `readOldTypesBundle` function

**Review by dzhelezov — APPROVED** [review]

File: substrate-metadata/src/util.ts [hunk 2]
```diff
@@ -210,3 +211,41 @@ export function isPreV14(metadata: Metadata): boolean {
...
+        res[specName] = {
+            types: {},
+            typesAlias: spec.alias,
+            signedExtensions: spec.signedExtensions,
+            versions: spec.types.map(v => ({
+                minmax: [
+                    v.minmax[0] || null,
+                    v.minmax[1] || null
+                ],
+                types: v.types,
```
Comment (eldargab) [inline substrate-metadata/src/util.ts:245]: Note, that we don't support the following polkadotjs features:

* [fallbacks](https://github.com/polkadot-js/api/blob/b24e13f478418ea7447b7c02eb3bc22b7e6a5452/packages/types/src/types/definitions.ts#L13)
* [aliases](https://github.com/polkadot-js/api/blob/b24e13f478418ea7447b7c02eb3bc22b7e6a5452/packages/types/src/types/definitions.ts#L13)

We need to throw an error when such features are used.

File: substrate-metadata/src/util.ts [hunk 3]
```diff
@@ -210,3 +211,41 @@ export function isPreV14(metadata: Metadata): boolean {
...
+
+export function convertPolkadotJSTypesBundle(typesByndle: PolkadotJSTypesBundle): Record<string, OldTypesBundle> {
+    let res: Record<string, OldTypesBundle> = {}
+
+    for (let specName in typesByndle.spec) {
+        let spec = typesByndle.spec[specName]
+        res[specName] = {
+            types: {},
+            typesAlias: spec.alias,
+            signedExtensions: spec.signedExtensions,
```
Comment (eldargab) [inline substrate-metadata/src/util.ts:239]: Our signed extensions are different.

File: substrate-processor/src/processor/handlerProcessor.ts [hunk 4]
```diff
@@ -622,7 +622,14 @@ export class SubstrateProcessor<Store> {
     }
 
     private getTypesBundle(specName: string, specVersion: number): OldTypesBundle {
```
Comment (eldargab) [inline substrate-processor/src/processor/handlerProcessor.ts:624]: This method should be modified to return `OldTypes`

**Review by eldargab — CHANGES_REQUESTED** [review]

File: substrate-ingest/src/main.ts [hunk 5]
```diff
@@ -25,6 +25,7 @@ runProgram(async () => {
     program.requiredOption('-e, --endpoint <url...>', 'WS rpc endpoint')
     program.option('-c, --endpoint-capacity <number...>', 'Maximum number of pending requests allowed for endpoint')
     program.option('--types-bundle <file>', 'JSON file with custom type definitions')
+    program.addOption(new Option('--format <name>', 'The the type bundle file format.').choices(['subsquid', 'polkadotjs']))
```
[the remark this answers is not in the document]
Comment (belopash) [reply substrate-ingest/src/main.ts:28]: Do we need it anywhere? I added such option to all tools that uses typesBundles

[the remark this answers is not in the document]
Comment (eldargab) [reply substrate-ingest/src/main.ts:28]: No, we don't.

File: substrate-metadata/src/util.ts (hunk 2 above)
Comment (belopash) [reply substrate-metadata/src/util.ts:245]: Maybe just warn?

Comment (eldargab) [reply substrate-metadata/src/util.ts:245]: Ok. Yes, both for fallback and aliases it is enough to warn.

File: substrate-metadata/src/util.ts (hunk 3 above)
Comment (belopash) [reply substrate-metadata/src/util.ts:239]: @eldargab should I convert polkadot.js signedExtensions to ours?

**Comment by eldargab** [comment]
Closing in favour of https://github.com/subsquid/squid/commit/a4718e1c2ec271a0f8159d8e8c1d296f435ad634

**Pull request closed without merging**